tiledb/cloud/dag/dag.py

```diff
@@ -79,8 +121,10 @@ def depends_on(self, node):
 
     def cancel(self):
         self.status = st.Status.CANCELLED
-        if self.future is not None:
-            self.future.cancel()
+        with self._future_lock:
```

Comment:
I think before the introduction of retry logic, this made sense, since the only invariant `_future_lock` was that both `_real_future` and `future` were both set up together and consistently. There was also a time where `_future_lock` was a more comprehensive lifecycle lock, and some of the operations that the callback would have called themselves would have acquired the `_future_lock`, which meant that we needed to release it (indicating that the invariants asserted by the lock held).

However, post–retry logic, that was no longer true and we should have indeed held onto this lock while calling `rft.cancel()`, because otherwise we might swap out the `_real_future` before this could happen.

That said, this entire thing is obsolete because it’s all much simpler now.